### DISCUSSION TITLE
How do we add new custom domains post-install ?

### DISCUSSION THREAD

**Discussion by zeroav-org** [simple-login/app#1060] [Self-hosting question]
Hi,

Set up app-ci:latest yesterday. Going well so far.

I have some custom domains in simplelogin.env since install:

`OTHER_ALIAS_DOMAINS=["domain32.com", "domain32.net"]`

These perform as expected.

Now I wanted to add domain32.org to the running install. 

I edited simplelogin.env and made it

`OTHER_ALIAS_DOMAINS=["domain32.com", "domain32.net", "domain32.org"]`

I then run 

`docker run --rm     --name sl-init     -v $(pwd)/sl:/sl     -v $(pwd)/simplelogin.env:/code/.env     -v $(pwd)/dkim.key:/dkim.key     -v $(pwd)/dkim.pub.key:/dkim.pub.key     --network="sl-network"     simplelogin/app-ci:latest python init_app.py`

It confirms that it added domain32.org. Looking good!

I restart sl-email container and the postfix service. Now e-mails to domain32.org bounce. 

When I check if simplelogin recognizes domain32.org as an SL domain it fails as described in point B1 [here](https://github.com/simple-login/app/blob/master/docs/troubleshooting.md). 

`postmap -q domain32.org pgsql:/etc/postfix/pgsql-relay-domains.cf`

returns nothing.

Full reboot made no difference. The domains I have since initial install work, the custom added later does not. DNS for domain32.org is correct. The new custom domain shows in the app and I can create aliases but when an e-mail is sent to the alias, SL will reject it

i.e. I create foobar@domain32.org and it succeeds but mails will bounce

```
554 5.7.1
    <[foobar@domain32.org](mailto:foobar@domain32.org)>: Relay access denied
```

Am I missing a step ? What is the process to add a new custom domain?

Edit: Running app-ci:latest  /git returns "5c5bafea18e85fad41da6094d40bd112993e4f87"

**Comment by zeroav-org** [comment]
Have updated to "9cf2f44166e376772a193330a3293ffa0f24b8e5" from ci-latest - same thing.

**Reply by nguyenkims** [reply]
@zeroav-org when modifying OTHER_ALIAS_DOMAINS,  we need to re-run the `init_app.py`, you can run it by:

```
sudo docker run --rm \
    --name sl-init \
    -v $(pwd)/sl:/sl \
    -v $(pwd)/sl/upload:/code/static/upload \
    -v $(pwd)/simplelogin.env:/code/.env \
    -v $(pwd)/dkim.key:/dkim.key \
    -v $(pwd)/dkim.pub.key:/dkim.pub.key \
    --network="sl-network" \
    simplelogin/app:3.4.0 python init_app.py
```

**Reply by zeroav-org** [reply]
> @zeroav-org when modifying OTHER_ALIAS_DOMAINS, we need to re-run the `init_app.py`, you can run it by:
> 
> ```
> sudo docker run --rm \
>     --name sl-init \
>     -v $(pwd)/sl:/sl \
>     -v $(pwd)/sl/upload:/code/static/upload \
>     -v $(pwd)/simplelogin.env:/code/.env \
>     -v $(pwd)/dkim.key:/dkim.key \
>     -v $(pwd)/dkim.pub.key:/dkim.pub.key \
>     --network="sl-network" \
>     simplelogin/app:3.4.0 python init_app.py
> ```

Oh, I just ran init like from initial install:

`docker run --rm --name sl-init -v $(pwd)/sl:/sl -v $(pwd)/simplelogin.env:/code/.env -v $(pwd)/dkim.key:/dkim.key -v $(pwd)/dkim.pub.key:/dkim.pub.key --network="sl-network" simplelogin/app-ci:latest python init_app.py`

So thats the issue. Will test and report back. Let me take this opportunity to thank you. We have bought subscriptions but will self-host anyway.

Edit: Tried:

```
sudo docker run --rm \
    --name sl-init \
    -v $(pwd)/sl:/sl \
    -v $(pwd)/sl/upload:/code/static/upload \
    -v $(pwd)/simplelogin.env:/code/.env \
    -v $(pwd)/dkim.key:/dkim.key \
    -v $(pwd)/dkim.pub.key:/dkim.pub.key \
    --network="sl-network" \
    simplelogin/app-ci:latest python init_app.py
```

Then restarted all containers. Emails to the new domain still bounce.

**Reply by nguyenkims** [reply]
that's indeed strange, can you check if the new domain is listed in the `public_domain` table in the database?

**Comment by zeroav-org** [comment]
`postmap -q domain32.org pgsql:/etc/postfix/pgsql-relay-domains.cf`

returns nothing.

I will inspect the db itself and see whats in that table and report back. Monday.

**Comment by zeroav-org** [comment]
> that's indeed strange, can you check if the new domain is listed in the `public_domain` table in the database?

OK I am back in the saddle. Set up pgadmin4 to make things easier. 

I can see that the added domain has been added to the `public_domain` table and has the same sort of flags set as all the others. `premium_only` and `can_use_subdomain` are false. The time stamp under `created_at` correctly reflects that it was added a day after initial install. It seems the database part of sl-init is working correctly but then postfix isn't told of the change.

**Reply by nguyenkims** [reply]
I see the issue now. Can you update the `/etc/postfix/pgsql-relay-domains.cf` to 

```
# postgres config
hosts = localhost
user = myuser
password = mypassword
dbname = simplelogin

query = SELECT domain FROM custom_domain WHERE domain='%s' AND verified=true
    UNION SELECT domain FROM public_domain WHERE domain='%s'
    UNION SELECT '%s' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

and `/etc/postfix/pgsql-transport-maps.cf` to

```
# postgres config
hosts = localhost
user = myuser
password = mypassword
dbname = simplelogin

# forward to smtp:127.0.0.1:20381 for custom domain AND email domain
query = SELECT 'smtp:127.0.0.1:20381' FROM custom_domain WHERE domain = '%s' AND verified=true
    UNION SELECT 'smtp:127.0.0.1:2025' FROM public_domain WHERE domain = '%s'
    UNION SELECT 'smtp:127.0.0.1:20381' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

Make sure to replace `mydomain.com` by your actual domain.

**Comment by zeroav-org** [comment]
This fixed the relaying issue! E-mails to the new custom domain do not bounce. But they are not delivered to the destination. In postfix' /var/log/mail.log I see:

`...relay=none, delay=729, delays=728/0.58/0/0, dsn=4.4.1, status=deferred (delivery temporarily suspended: connect to 127.0.0.1[127.0.0.1]:2025: Connection refused)`

But e-mails to aliases of the base domain work fine.

**Reply by nguyenkims** [reply]
ah sorry, my mistake, the `/etc/postfix/pgsql-transport-maps.cf` should rather be

```
# postgres config
hosts = localhost
user = myuser
password = mypassword
dbname = simplelogin

# forward to smtp:127.0.0.1:20381 for custom domain AND email domain
query = SELECT 'smtp:127.0.0.1:20381' FROM custom_domain WHERE domain = '%s' AND verified=true
    UNION SELECT 'smtp:127.0.0.1:20381' FROM public_domain WHERE domain = '%s'
    UNION SELECT 'smtp:127.0.0.1:20381' WHERE '%s' = 'mydomain.com' LIMIT 1;
```

**Reply by zeroav-org** [reply]
SOLVED!!!! 
Just updated to `39aeb81f9a87e86ccd67590e7f1c56e9874768e1 `and all seems to be working now! Thank you!
